### PR TITLE
Fix create partition of an exist topic doesn't throw RestException

### PR DESCRIPTION
**Pull request by BewareMyPower**
### Motivation

Currently creating a partition of an existed partitioned topic was created doesn't throw any exception. However it should be an invalid behavior. The reason is that when a non partitioned topic was created, it only checks whether the number of partitions is positive. However, no matter the topic doesn't exist or the topic is an existed partition, the number of partitions is 0. This PR is to distinguish these two cases and throw a `RestException` when the non-partitioned topic is an existed partition.

### Modifications

- When creating a non-partition topic that is an existed partition, throw a `RestException` and add a test to verify it.
- Add a test provider to ensure backward compatibility that users can still create missed partitions by creating non-partitioned topics of the missed partition.
- Fix the functions worker startup. Before this PR it always tries to create the metadata topics like `public/functions/assignment` no matter if it exists. Here we ignore the `ConflictException`.

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

This change added tests and can be verified as follows:

  - *Added PersistentTopics#testCreateExistedPartition*
  - *Added another test provider to PartitionCreationTest#testCreateMissedPartitions*
  - *Add tests for creating existed topics to `AdminApiTeststestPersistentTopicCreation`.*

### CODE REVIEW

**Comment by BewareMyPower** [comment]
@hangc0276 PTAL, this fix will expose the wrong repeated metadata initialization code of KoP.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 1]
```diff
@@ -460,6 +460,16 @@ protected void internalCreateNonPartitionedTopic(boolean authoritative) {
         if (partitionMetadata.partitions > 0) {
             log.warn("[{}] Partitioned topic with the same name already exists {}", clientAppId(), topicName);
             throw new RestException(Status.CONFLICT, "This topic already exists");
+        } else {
+            if (topicName.isPartitioned()) {
+                final TopicName partitionedTopicName = TopicName.get(topicName.getPartitionedTopicName());
+                partitionMetadata = fetchPartitionedTopicMetadata(pulsar(), partitionedTopicName);
```
Comment (zymap) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:466]: I think we shouldn't allow creating a topic with the partitioned topic name style even if the partition is not existing. That would be weird there has a non-partitioned topic with a partitioned topic name when you listing topics.  We can improve this check in the `validateNonPartitionTopicName`. What do you think?

Comment (BewareMyPower) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:466]: Sounds good. I'll use the suggested way to fix.

Comment (BewareMyPower) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:466]: After rethinking for a while, I think a better way is to check if `/managed-ledgers/<tenant>/<namespace>/<topic>` exists.

No matter if disallow creating `xxx-partition-i` or the way I adopted, the compatibility will be broken. Because if users upgrade their pulsar from a low version, they may rely on creating a `xxx-partition-0` to create missing partitions instead of the REST API, as well as some old code of other clients.

**Comment by BewareMyPower** [comment]
I've changed another way to fix the issue and the backward compatibility could be retained, PTAL again, @zymap

**Comment by zymap** [comment]
> By the way, when a AdminResource was created with a full topic name, the internal topicName would be initialized with a wrong name. eg for persistent://public/default/my-topic, the internal topicName is persistent://public /default/persistent://public/default/my-topic. This wrong behavior should be fixed.

I am curious about this, when will this behavior happen? As I know, the rest API is defined as `tenant/namespace/topicname`. So that means this case will happen when you name your topic name as `persistent://public /default/topic` explicitly.

> Currently creating a partition of an existed partitioned topic was created doesn't throw any exception. However it should be an invalid behavior. The reason is that when a non partitioned topic was created, it only checks whether the number of partitions is positive. However, no matter the topic doesn't exist or the topic is an existed partition, the number of partitions is 0. This PR is to distinguish these two cases and throw a RestException when the non-partitioned topic is an existed partition.

About this issue, I am using master code to test and it shows I can't create a non-partitioned topic with a name that contains the `-partition-`

```
 bin/pulsar-admin topics create persistent://public/default/tttest-partition-0
08:46:49.633 [AsyncHttpClient-7-1] WARN  org.apache.pulsar.client.admin.internal.BaseResource - [http://localhost:8080/admin/v2/persistent/public/default/tttest-partition-0] Failed to perform http put request: javax.ws.rs.ClientErrorException: HTTP 412 Precondition Failed
Can't create topic tttest-partition-0 with "-partition-" followed by numeric value if there isn't a partitioned topic tttest created.

Reason: Can't create topic tttest-partition-0 with "-partition-" followed by numeric value if there isn't a partitioned topic tttest created.
```

So could you please give me more information about this case?

**Comment by BewareMyPower** [comment]
@zymap You're right for the 1st issue. It only affects my tests code:

```java
        final String partitionName = TopicName.get(topicName).getPartition(0).toString();
        try {
            persistentTopics.createNonPartitionedTopic(testTenant, testNamespace, partitionName, false);
```

I should use the local name in tests instead of in the `validateTopicName`.

For the 2st issue, the intention for creating non-partitioned topic with `-partition-` suffix is to create missed partitions. So you should first create a partitioned topic.

Here's an example:

```bash
# 1. A partition was somehow deleted or not created successfully
$ ./bin/pulsar-admin topics create-partitioned-topic xyz -p 3
$ ./bin/pulsar-admin topics delete xyz-partition-0
$ ./bin/pulsar-admin topics list public/default              
"persistent://public/default/xyz-partition-2"
"persistent://public/default/xyz-partition-1"
# 2. Then we can create a non-partitioned topic to create missed partitions
$ ./bin/pulsar-admin topics create xyz-partition-0           
$ ./bin/pulsar-admin topics list public/default   
"persistent://public/default/xyz-partition-2"
"persistent://public/default/xyz-partition-0"
"persistent://public/default/xyz-partition-1"
# 3. However, if you create a existed partition, there's no error, this is what I want to fix
$ ./bin/pulsar-admin topics create xyz-partition-2
```

The rule is from https://github.com/apache/pulsar/pull/5148.

**Comment by BewareMyPower** [comment]
The test is now broken, I'll fix it first.

**Review by codelipenghui — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/AdminResource.java [hunk 2]
```diff
@@ -356,8 +356,6 @@ protected void validateTopicName(String property, String namespace, String encod
                     topic, e);
             throw new RestException(Status.PRECONDITION_FAILED, "Topic name is not valid");
         }
-
-        this.topicName = TopicName.get(domain(), namespaceName, topic);
```
Comment (zymap) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/AdminResource.java:360]: Unrelated changes?

**Review by zymap — COMMENTED** [review]
I see. Thanks for your explanation!

Left a minor comment.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/AdminResource.java (hunk 2 above)
Comment (BewareMyPower) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/AdminResource.java:360]: Yes, it's a repeated code (see line 353). I've modified this method before, when I changed it back, I kept this line removed.

**Review by zymap — APPROVED** [review]

**Review by sijie — APPROVED** [review]

**Comment by BewareMyPower** [comment]
It looks like the `BrokerServiceAutoSubscriptionCreationTest` is affected by this PR. I'll fix it soon.

**Review by jiazhai — APPROVED** [review]

**Comment by BewareMyPower** [comment]
The failed tests may be caused by this PR, I'm trying to fix them.

**Comment by BewareMyPower** [comment]
I ran the functions integration tests in my local environment and got the similar error:

> 19:57:06.676 [docker-java-stream-1645855842:org.apache.pulsar.tests.integration.utils.DockerUtils$2@221] INFO  org.apache.pulsar.tests.integration.utils.DockerUtils - DOCKER.exec(PulsarFunctionsProcessTest-apuqq-pulsar-broker-1:tail -f /var/log/pulsar/broker.log): STDOUT: 11:57:06.237 [pulsar-web-43-7] ERROR org.apache.pulsar.broker.admin.impl.PersistentTopicsBase - [null] Topic persistent://public/functions/assignments already exists
> 11:57:06.246 [pulsar-web-43-7] INFO  org.eclipse.jetty.server.RequestLog - 172.28.0.10 - - [03/Feb/2021:11:57:06 +0000] "PUT /admin/v2/persistent/public/functions/assignments?authoritative=true HTTP/1.1" 409 38 "-" "Pulsar-Java-v2.8.0-SNAPSHOT" 18

It looks like the functions worker container tries to create existed `public/functions/assignments`, which leads to the failure of test setup.

**Comment by BewareMyPower** [comment]
There're still 3 tests that failed.

> [ERROR]   Run 17: PulsarFunctionsProcessTest.testSerdeFunction » ServerSideError HTTP 502 Bad Ga...
> [ERROR]   Run 19: PulsarFunctionsProcessTest.testSlidingCountWindowTest » ServerSideError HTTP 5...
> [ERROR]   Run 21: PulsarFunctionsProcessTest.testTumblingCountWindowTest » ServerSideError HTTP ...

**Comment by BewareMyPower** [comment]
Now all tests passed, PTAL again, @jiazhai @sijie @zymap @codelipenghui
